tests/socket_tests.rb

```diff
@@ -0,0 +1,114 @@
+# A fake Excon::Socket that allows passing in an arbitrary backend @socket
+class MockExconSocket < Excon::Socket
+  attr_reader :read_count
+
+  def initialize(backend_socket, *args)
+    super(*args)
+    @read_count = 0
+    @socket = backend_socket
+  end
+
+  def read_nonblock(*args)
+    @read_count += 1
+    super
+  end
+
+  def connect
+    # pass
+  end
+
+  def select_with_timeout(*args)
+    # don't actually wait, assume we're ready
+  end
+end
+
+# A socket whose read_nonblock returns from an input list,
+# and which counts the number of reads
+class MockNonblockRubySocket
+  attr_reader :sequence
+
+  def initialize(nonblock_reads)
+    @nonblock_reads = nonblock_reads
+    @sequence = []
+  end
+
+  def read_nonblock(maxlen)
+    if @nonblock_reads.empty?
+      @sequence << 'EOF'
+      raise EOFError
+    elsif @nonblock_reads.first.empty?
+      @nonblock_reads.shift
+      if @nonblock_reads.empty?
+        @sequence << 'EOF'
+        raise EOFError
+      end
+      @sequence << 'EAGAIN'
+      raise Errno::EAGAIN
+    elsif
+      len = maxlen ? maxlen : @nonblock_reads.first.length
+      ret = @nonblock_reads.first.slice!(0, len)
+      @sequence << ret.length
+      ret
+    end
+  end
+
+  # Returns the results of `block`, as well as how many times we called read on the Excon
+  # socket, and the sequence of reads on the backend socket
+  def self.check_reads(nonblock_reads, socket_args, &block)
+    backend_socket = MockNonblockRubySocket.new(nonblock_reads)
+    socket = MockExconSocket.new(backend_socket, { nonblock: true }.merge(socket_args))
+    ret = block[socket]
+    [ret, socket.read_count, backend_socket.sequence]
+  end
+end
+
+Shindo.tests('socket') do
```

Comment:
From what you describe, I imagine a number of these tests may have been failing at the outset of this. Could you describe what you were seeing prior to the changes you made to help me have context? Thanks!

Reply:
Will take a look at this soon.

Reply:
Thanks. I just wanted to have a sense of what kind of improvement you've seen and also what to lookout for in terms of potential regressions in the future.